# **PROYECTO INDIVIDUAL MLOps**

### Machine Learning Operations 

### **1. IMPORTACIÓN DE LIBRERÍAS**

In [2]:
import pandas as pd
import numpy as np 
import json
import glob 
import ast 
import gzip
import re
import pyarrow as pa
import pyarrow.parquet as pq

%load_ext autoreload
%autoreload 2

1.1 **CONVERTIR DATOS ANIDADOS DE UN JSON EN UN DATAFRAME CON PANDAS EN PYTHON**

- Creamos una lista vacía llamada "v_general" donde vamos a almacenar nuestros datos.
- Abrimos un archivo comprimido en formato GZIP llamado "steam_games.json.gz".
- Lo abrimos en modo de lectura de texto ('rt') y especificamos que el texto está codificado en UTF-8.
- Recorremos el archivo línea por línea.
- En cada línea, interpretamos el contenido como datos en formato JSON.
- Luego, agregamos esos datos a la lista 'v_general'.

In [3]:
v_general = []

with gzip.open("C:\\Users\\ACER\\Documents\\HENRY\\LABS\\PI01\\DATASETS\\steam_games.json.gz", 'rt', encoding='utf-8') as file:
    for line in file:
        data = json.loads(line)
        v_general.append(data)

Se convierte en DataFrame

In [4]:
Games = pd.DataFrame(v_general)
Games

publisher                                 genres   
0                   NaN                                    NaN  \
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title   
0                            NaN                       NaN  \
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date   
0                                                     NaN          NaN  \
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags   
0                                                     NaN  \
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  discount_price   
0                                                     NaN             NaN  \
1                                                     NaN             NaN   
2                                                     NaN             NaN   
3                                                     NaN             NaN   
4                                                     NaN             NaN   
...                                                   ...             ...   
120440  http://steamcommunity.com/app/773640/reviews/?...            1.49   
120441  http://steamcommunity.com/app/733530/reviews/?...            4.24   
120442  http://steamcommunity.com/app/610660/reviews/?...            1.39   
120443  http://steamcommunity.com/app/658870/reviews/?...             NaN   
120444 

**EXPLORACIÓN DE DATA**

In [5]:
Games.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'discount_price', 'specs', 'price',
       'early_access', 'id', 'metascore', 'developer', 'user_id', 'steam_id',
       'items', 'items_count'],
      dtype='object')

In [6]:
Games.dtypes

publisher          object
genres             object
app_name           object
title              object
url                object
release_date       object
tags               object
reviews_url        object
discount_price    float64
specs              object
price              object
early_access       object
id                 object
metascore          object
developer          object
user_id            object
steam_id           object
items              object
items_count       float64
dtype: object

In [7]:
Games.describe()

discount_price   items_count
count      225.000000  88310.000000
mean        11.930533     58.353629
std         17.492643    122.312095
min          0.490000      0.000000
25%          1.390000      3.000000
50%          4.190000     26.000000
75%         22.660000     73.000000
max        139.990000   7762.000000

In [8]:
Games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   publisher       24083 non-null  object 
 1   genres          28852 non-null  object 
 2   app_name        32133 non-null  object 
 3   title           30085 non-null  object 
 4   url             32135 non-null  object 
 5   release_date    30068 non-null  object 
 6   tags            31972 non-null  object 
 7   reviews_url     32133 non-null  object 
 8   discount_price  225 non-null    float64
 9   specs           31465 non-null  object 
 10  price           30758 non-null  object 
 11  early_access    32135 non-null  object 
 12  id              32133 non-null  object 
 13  metascore       2677 non-null   object 
 14  developer       28836 non-null  object 
 15  user_id         88310 non-null  object 
 16  steam_id        88310 non-null  object 
 17  items           88310 non-nul

In [9]:
Games.shape

(120445, 19)

**SE ELIMINAN LAS SIGUIENTES COLUMNAS YA QUE SE CUENTA CON DICHA INFORMACIÓN EN EL DATASET DE ITEMS**

In [10]:
Games = Games.drop(columns = ['items', 'items_count', 'steam_id','user_id',], axis=1)
Games

publisher                                 genres   
0                   NaN                                    NaN  \
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title   
0                            NaN                       NaN  \
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date   
0                                                     NaN          NaN  \
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags   
0                                                     NaN  \
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  discount_price   
0                                                     NaN             NaN  \
1                                                     NaN             NaN   
2                                                     NaN             NaN   
3                                                     NaN             NaN   
4                                                     NaN             NaN   
...                                                   ...             ...   
120440  http://steamcommunity.com/app/773640/reviews/?...            1.49   
120441  http://steamcommunity.com/app/733530/reviews/?...            4.24   
120442  http://steamcommunity.com/app/610660/reviews/?...            1.39   
120443  http://steamcommunity.com/app/658870/reviews/?...             NaN   
120444 

VERIFICACIÓN DE DATOS NULOS 


In [11]:
#Se verifica la cantidad de datos nulos que hay por columna y se suman los datos 
nulls = Games.isnull().sum()
nulls

publisher          96362
genres             91593
app_name           88312
title              90360
url                88310
release_date       90377
tags               88473
reviews_url        88312
discount_price    120220
specs              88980
price              89687
early_access       88310
id                 88312
metascore         117768
developer          91609
dtype: int64

In [12]:
Final_nulls = Games.isnull().sum().sum()
Final_nulls

1406985

Se eliminan los datos nulos encontrados 


In [13]:
SteamGames = Games.dropna(how='all').reset_index(drop=True)
SteamGames

publisher                                             genres   
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]  \
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title   
0           Lost Summoner Kitty       Lost Summoner Kitty  \
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags   
0          [Strategy, Action, Indie, Casual, Simulation]  \
1      [Free to Play, Strategy, Indie, RPG, Card Game...   
2      [Free to Play, Simulation, Sports, Casual, Ind...   
3                            [Action, Adventure, Casual]   
4                        [Action, Indie, Casual, Sports]   
...                                                  ...   
32130              [Strategy, Indie, Casual, Simulation]   
32131                          [Strategy, Indie, Casual]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
1      http://steamcommunity.com/app/643980/reviews/?...             NaN   
2      http://steamcommunity.com/app/670290/reviews/?...             NaN   
3      http://steamcommunity.com/app/767400/reviews/?...            0.83   
4      http://steamcommunity.com/app/773570/reviews/?...            1.79   
...                                                  ...             ...   
32130  http://steamcommunity.com/app/773640/reviews/?...            1.49   
32131  http://steamcommunity.com/app/733530/reviews/?...            4.24   
32132  http://steamcommunity.com/app/610660/reviews/?...      

In [14]:
#Se buscan y analizan datos duplicados 
Verificar = ['id']
Col_dups = SteamGames[SteamGames.duplicated(subset=Verificar)]
Col_dups

publisher               genres   
14573                                 Bethesda Softworks             [Action]  \
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name   
14573                Wolfenstein II: The New Colossus  \
30961  Batman: Arkham City - Game of the Year Edition   

                                                title   
14573                Wolfenstein II: The New Colossus  \
30961  Batman: Arkham City - Game of the Year Edition   

                                                     url release_date   
14573  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26  \
30961           http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags   
14573  [Action, FPS, Gore, Violent, Alternate History...  \
30961  [Action, Open World, Batman, Adventure, Stealt...   

                                             reviews_url  discount_price   
14573  http://steamcommunity.com/app/612880/reviews/?...             NaN  \
30961                                                NaN             NaN   

                                                   specs  price early_access   
14573  [Single-player, Steam Achievements, Full contr...  59.99        False  \
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

           id metascore                                   developer  
14573  612880        86                               Machine Games  
30961     NaN        91  Rocksteady Studios,Feral Interactive (Mac)

Procedemos a revisar los datos duplicados

In [15]:
SteamGames[SteamGames["developer"]== "Machine Games"]

publisher    genres   
12237  Bethesda Softworks  [Action]  \
13318  Bethesda Softworks  [Action]   
13894  Bethesda Softworks  [Action]   
14573  Bethesda Softworks  [Action]   
28996  Bethesda Softworks  [Action]   

                                                app_name   
12237  Wolfenstein II: The Freedom Chronicles - Seaso...  \
13318  Wolfenstein II: The Freedom Chronicles - Episo...   
13894                   Wolfenstein II: The New Colossus   
14573                   Wolfenstein II: The New Colossus   
28996                         Wolfenstein: The New Order   

                                                   title   
12237  Wolfenstein II: The Freedom Chronicles - Seaso...  \
13318  Wolfenstein II: The Freedom Chronicles - Episo...   
13894                   Wolfenstein II: The New Colossus   
14573                   Wolfenstein II: The New Colossus   
28996                         Wolfenstein: The New Order   

                                                     url release_date   
12237  http://store.steampowered.com/app/650410/Wolfe...   2017-11-07  \
13318  http://store.steampowered.com/app/624620/Wolfe...   2017-12-13   
13894          http://store.steampowered.com/app/612880/   2017-10-26   
14573  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26   
28996  http://store.steampowered.com/app/201810/Wolfe...   2014-05-19   

                                                    tags   
12237                            [Gore, Action, Violent]  \
13318                                  [Action, Violent]   
13894  [Action, FPS, Gore, Violent, Alternate History...   
14573  [Action, FPS, Gore, Violent, Alternate History...   
28996  [FPS, Action, Shooter, Alternate History, Firs...   

                                             reviews_url  discount_price   
12237  http://steamcommunity.com/app/650410/reviews/?...             NaN  \
13318  http://steamcommunity.com/app/624620/reviews/?...             NaN   
13894  http://steamcommunity.com/app/612880/reviews/?...             NaN   
14573  http://steamcommunity.com/app/612880/reviews/?...             NaN   
28996  http://steamcommunity.com/app/201810/reviews/?...             NaN   

                                                   specs  price early_access   
12237  [Single-player, Downloadable Content, Full con...  24.99        False  \
13318  [Single-player, Downloadable Content, Steam Ac...   9.99        False   
13894  [Single-player, Steam Achievements, Full contr...  59.99        False   
14573  [Single-player, Steam Achievements, Full contr...  59.99        False   
28996  [Single-player, Steam Achievements, Partial Co...  19.99        False   

           id metascore      developer  
12237  650410       NaN  Machine Games  
13318  624620       NaN  Machine Games  
13894  612880        86  Machine Games  
14573  612880        86  Machine Games  
28996  201810        81  Machine Games

In [16]:
SteamGames[SteamGames["developer"]== "Rocksteady Studios,Feral Interactive (Mac)"]

publisher               genres   
1068   Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]  \
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   
31617  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name   
1068   Batman: Arkham City - Game of the Year Edition  \
30961  Batman: Arkham City - Game of the Year Edition   
31617  Batman: Arkham Asylum Game of the Year Edition   

                                                title   
1068   Batman: Arkham City - Game of the Year Edition  \
30961  Batman: Arkham City - Game of the Year Edition   
31617  Batman: Arkham Asylum Game of the Year Edition   

                                                     url release_date   
1068   http://store.steampowered.com/app/200260/Batma...   2012-09-07  \
30961           http://store.steampowered.com/app/200260   2012-09-07   
31617  http://store.steampowered.com/app/35140/Batman...   2010-03-26   

                                                    tags   
1068   [Action, Open World, Batman, Adventure, Stealt...  \
30961  [Action, Open World, Batman, Adventure, Stealt...   
31617  [Action, Batman, Stealth, Adventure, Third Per...   

                                             reviews_url  discount_price   
1068   http://steamcommunity.com/app/200260/reviews/?...             NaN  \
30961                                                NaN             NaN   
31617  http://steamcommunity.com/app/35140/reviews/?b...             NaN   

                                                   specs  price early_access   
1068   [Single-player, Steam Achievements, Steam Trad...  19.99        False  \
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   
31617  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

           id metascore                                   developer  
1068   200260        91  Rocksteady Studios,Feral Interactive (Mac)  
30961     NaN        91  Rocksteady Studios,Feral Interactive (Mac)  
31617   35140        91  Rocksteady Studios,Feral Interactive (Mac)

Al revisar los datos notamos que la fila 1068 y la 30961 cuentan con datos identicos excepto que la 30961 cuenta con algunos datos nulos, por ellos se procede a eliminarla. 
Asimismo las filas 14573 y 13894 son identicas así que podemos proceder a eliminar cualquiera de ellas. 

In [17]:
#elimino columnas con id duplicadas
Del_dups = [14573, 30961]
SteamGames= SteamGames.drop(Del_dups)
SteamGames

publisher                                             genres   
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]  \
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title   
0           Lost Summoner Kitty       Lost Summoner Kitty  \
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags   
0          [Strategy, Action, Indie, Casual, Simulation]  \
1      [Free to Play, Strategy, Indie, RPG, Card Game...   
2      [Free to Play, Simulation, Sports, Casual, Ind...   
3                            [Action, Adventure, Casual]   
4                        [Action, Indie, Casual, Sports]   
...                                                  ...   
32130              [Strategy, Indie, Casual, Simulation]   
32131                          [Strategy, Indie, Casual]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
1      http://steamcommunity.com/app/643980/reviews/?...             NaN   
2      http://steamcommunity.com/app/670290/reviews/?...             NaN   
3      http://steamcommunity.com/app/767400/reviews/?...            0.83   
4      http://steamcommunity.com/app/773570/reviews/?...            1.79   
...                                                  ...             ...   
32130  http://steamcommunity.com/app/773640/reviews/?...            1.49   
32131  http://steamcommunity.com/app/733530/reviews/?...            4.24   
32132  http://steamcommunity.com/app/610660/reviews/?...      

In [18]:
SteamGames.shape

(32133, 15)

#### EL dataset aún es un archivo json, por lo que el tipo de dato object , lo que puede ser datos anidados con formato clave - valor. Entonces a través de un bucle se analiza columna por columna y se almacenan en un diccionario vacio para luego transformarlo en un dataset y alli poder ver el tipo de datos de las columnas.

In [19]:
DataType = {"columna":[],"DataType":[]} #genero un diccionario vacio para ir almacenando lo que genere el bucle

for columna in SteamGames.columns: #un bucle que va recorriendo 
    DataType["columna"].append(columna)
    DataType["DataType"].append(SteamGames[columna].apply(type).unique())

Analyst= pd.DataFrame(DataType)
Analyst

columna                                         DataType
0        publisher                 [<class 'str'>, <class 'float'>]
1           genres                [<class 'list'>, <class 'float'>]
2         app_name                 [<class 'str'>, <class 'float'>]
3            title                 [<class 'str'>, <class 'float'>]
4              url                                  [<class 'str'>]
5     release_date                 [<class 'str'>, <class 'float'>]
6             tags                [<class 'list'>, <class 'float'>]
7      reviews_url                 [<class 'str'>, <class 'float'>]
8   discount_price                                [<class 'float'>]
9            specs                [<class 'list'>, <class 'float'>]
10           price                 [<class 'float'>, <class 'str'>]
11    early_access                                 [<class 'bool'>]
12              id                 [<class 'str'>, <class 'float'>]
13       metascore  [<class 'float'>, <class 'int'>, <class 'str'>]
14       developer                 [<class 'str'>, <class 'float'>]

### **Transformación de la columnas "price", "release_date" y "genres"**

Verificamos que los datos de la columna price sea númericos y si no lo son revisamos cuales deben ser reemplazados 

In [20]:
Games['price'].unique()

array([nan, 4.99, 'Free To Play', 'Free to Play', 0.99, 2.99, 3.99, 9.99,
       18.99, 29.99, 'Free', 10.99, 1.59, 14.99, 1.99, 59.99, 8.99, 6.99,
       7.99, 39.99, 19.99, 7.49, 12.99, 5.99, 2.49, 15.99, 1.25, 24.99,
       17.99, 61.99, 3.49, 11.99, 13.99, 'Free Demo', 'Play for Free!',
       34.99, 74.76, 1.49, 32.99, 99.99, 14.95, 69.99, 16.99, 79.99,
       49.99, 5.0, 44.99, 13.98, 29.96, 119.99, 109.99, 149.99, 771.71,
       'Install Now', 21.99, 89.99, 'Play WARMACHINE: Tactics Demo', 0.98,
       139.92, 4.29, 64.99, 'Free Mod', 54.99, 74.99, 'Install Theme',
       0.89, 'Third-party', 0.5, 'Play Now', 299.99, 1.29, 3.0, 15.0,
       5.49, 23.99, 49.0, 20.99, 10.93, 1.39, 'Free HITMAN™ Holiday Pack',
       36.99, 4.49, 2.0, 4.0, 9.0, 234.99, 1.95, 1.5, 199.0, 189.0, 6.66,
       27.99, 10.49, 129.99, 179.0, 26.99, 399.99, 31.99, 399.0, 20.0,
       40.0, 3.33, 199.99, 22.99, 320.0, 38.85, 71.7, 59.95, 995.0, 27.49,
       3.39, 6.0, 19.95, 499.99, 16.06, 4.68, 131.4, 44.

Se crea una función para poder reemplazar los datos de tipo str por tipo float 0.0

In [21]:
# Definir una función lambda que convierte el valor a float o devuelve 0.0 en caso de error
cambio_a_float = lambda valor: float(valor) if isinstance(valor, (float, int)) else 0.0 if pd.isna(valor) else 0.0

# Aplicar la función lambda a la columna precio
SteamGames["price"] = SteamGames["price"].apply(cambio_a_float)
SteamGames.info()


<class 'pandas.core.frame.DataFrame'>
Index: 32133 entries, 0 to 32134
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   publisher       24081 non-null  object 
 1   genres          28850 non-null  object 
 2   app_name        32131 non-null  object 
 3   title           30083 non-null  object 
 4   url             32133 non-null  object 
 5   release_date    30066 non-null  object 
 6   tags            31970 non-null  object 
 7   reviews_url     32132 non-null  object 
 8   discount_price  225 non-null    float64
 9   specs           31463 non-null  object 
 10  price           30756 non-null  float64
 11  early_access    32133 non-null  object 
 12  id              32132 non-null  object 
 13  metascore       2675 non-null   object 
 14  developer       28834 non-null  object 
dtypes: float64(2), object(13)
memory usage: 3.9+ MB


In [22]:
SteamGames["price"].value_counts()

price
4.99      4278
9.99      3902
2.99      3429
0.99      2607
1.99      2541
          ... 
26.99        1
179.00       1
10.49        1
6.66         1
160.91       1
Name: count, Length: 145, dtype: int64

Se buscan outliers en precio, se identifican los umbrales y se identifican los outliers 

In [23]:
price_description = SteamGames['price'].describe()
Q1 = price_description['25%']
Q3 = price_description['75%']
IQR = Q3 - Q1

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)

Q1: 2.99
Q3: 9.99
IQR: 7.0


In [24]:
Superior = SteamGames['price'].quantile(0.75) + 1.5 * (SteamGames['price'].quantile(0.75) - SteamGames['price'].quantile(0.25))
Inferior = SteamGames['price'].quantile(0.25) - 1.5 * (SteamGames['price'].quantile(0.75) - SteamGames['price'].quantile(0.25))

print("Umbral Superior:", Superior)
print("Umbral Inferior:", Inferior)

Umbral Superior: 20.490000000000002
Umbral Inferior: -7.51


In [25]:
outliers_sup = SteamGames.loc[SteamGames['price'] > Superior]
outliers_inf = SteamGames.loc[SteamGames['price'] < Inferior]

print(outliers_sup)
print(outliers_inf)

                              publisher   
8                Poppermost Productions  \
18               Poppermost Productions   
51                                  NaN   
59                               SimBin   
234                              NCsoft   
...                                 ...   
31938  Firedog Creative Company Limited   
31956                               NaN   
31959                 CarX Technologies   
32070                  DreamInCode B.V.   
32098                        Koch Media   

                                                  genres   
8              [Free to Play, Indie, Simulation, Sports]  \
18             [Free to Play, Indie, Simulation, Sports]   
51                                                   NaN   
59                                              [Racing]   
234                         [Massively Multiplayer, RPG]   
...                                                  ...   
31938                                  [RPG, Simulation]   
319

In [26]:
outliers_inf

Empty DataFrame
Columns: [publisher, genres, app_name, title, url, release_date, tags, reviews_url, discount_price, specs, price, early_access, id, metascore, developer]
Index: []

In [27]:
outliers_sup

publisher   
8                Poppermost Productions  \
18               Poppermost Productions   
51                                  NaN   
59                               SimBin   
234                              NCsoft   
...                                 ...   
31938  Firedog Creative Company Limited   
31956                               NaN   
31959                 CarX Technologies   
32070                  DreamInCode B.V.   
32098                        Koch Media   

                                                  genres   
8              [Free to Play, Indie, Simulation, Sports]  \
18             [Free to Play, Indie, Simulation, Sports]   
51                                                   NaN   
59                                              [Racing]   
234                         [Massively Multiplayer, RPG]   
...                                                  ...   
31938                                  [RPG, Simulation]   
31956                                                NaN   
31959  [Massively Multiplayer, Racing, Simulation, Sp...   
32070  [Action, Adventure, Casual, Indie, Simulation,...   
32098                      [Utilities, Video Production]   

                                             app_name   
8                       SNOW - All Access Legend Pass  \
18                               SNOW - Lifetime Pack   
51                   Medicalholodeck Personal Version   
59                               RACE - The WTCC Game   
234                                Guild Wars Trilogy   
...                                               ...   
31938                                           愛神餐館2   
31956  Three Kingdoms VR - Jade Knight (光之三國VR - 青龍騎)   
31959               CarX Drift Racing Online - Deluxe   
32070                                       Mesozoica   
32098                       CyberLink YouCam 7 Deluxe   

                                   title   
8          SNOW - All Access Legend Pass  \
18                  SNOW - Lifetime Pack   
51                                   NaN   
59                  RACE - The WTCC Game   
234                   Guild Wars Trilogy   
...                                  ...   
31938                              愛神餐館2   
31956                                NaN   
31959  CarX Drift Racing Online - Deluxe   
32070                          Mesozoica   
32098          CyberLink YouCam 7 Deluxe   

                                                     url release_date   
8      http://store.steampowered.com/app/774278/SNOW_...   2018-01-04  \
18     http://store.steampowered.com/app/774279/SNOW_...   2018-01-03   
51     http://store.steampowered.com/app/769350/Medic...          NaN   
59     http://store.steampowered.com/app/4230/RACE__T...   2006-11-24   
234    http://store.steampowered.com/app/29570/Guild_...   2008-09-25   
...                                                  ...          ...   
31938        http://store.steampowered.com/app/769680/2/   2017-12-29   
31956  http://store.steampowered.com/app/731120/Three...          NaN   
31959  http://store.steampowered.com/app/776950/CarX_...   2017-12-28   
32070  http://store.steampowered.com/app/719590/Mesoz...   2018-01-02   
32098  http://store.steampowered.com/app/755310/Cyber...   2017-12-26   

                                                    tags   
8              [Free to Play, Indie, Simulation, Sports]  \
18             [Free to Play, Indie, Simulation, Sports]   
51                                           [Education]   
59                                  [Racing, Simulation]   
234            [RPG, Massively Multiplayer, MMORPG, PvP]   
...                                                  ...   
31938                           [RPG, Simulation, Anime]   
31956               [Action, Casual, Simulation, Sports]   
31959  [Massively Multiplayer, Simulation, Racing, Sp...   
32070  [Early Access, Simulation, Strategy, Casual, I...   
32098                      [Utilities, Vi

RELEASE_DATE 


In [28]:
SteamGames["release_date"].value_counts()

release_date
2012-10-16    100
2017-08-31     92
2017-09-26     89
2017-06-21     82
2017-07-25     78
             ... 
1988-04-16      1
2013-08-24      1
2011-05-07      1
2010-08-21      1
2018-10-01      1
Name: count, Length: 3582, dtype: int64

In [29]:
SteamGames["release_date"]

0        2018-01-04
1        2018-01-04
2        2017-07-24
3        2017-12-07
4               NaN
            ...    
32130    2018-01-04
32131    2018-01-04
32132    2018-01-04
32133    2017-09-02
32134           NaN
Name: release_date, Length: 32133, dtype: object

In [30]:
# Convierte la columna 'release_date' a cadenas de texto (strings)
# Define una función lambda para extraer el año o asignar 'sin dato' en caso de que no haya un año válido
# Aplica la función lambda a la columna 'release_date' y asigna el resultado a 'Año_estreno'

SteamGames['release_date'] = SteamGames['release_date'].astype(str)
extract_year = lambda date: re.search(r'\d{4}', date).group() if re.search(r'\d{4}', date) else 'sin dato'
SteamGames['Año_estreno'] = SteamGames['release_date'].apply(extract_year)

GENRES 

In [31]:
SteamGames['genres']

0            [Action, Casual, Indie, Simulation, Strategy]
1                     [Free to Play, Indie, RPG, Strategy]
2        [Casual, Free to Play, Indie, Simulation, Sports]
3                              [Action, Adventure, Casual]
4                                                      NaN
                               ...                        
32130                [Casual, Indie, Simulation, Strategy]
32131                            [Casual, Indie, Strategy]
32132                          [Indie, Racing, Simulation]
32133                                      [Casual, Indie]
32134                                                  NaN
Name: genres, Length: 32133, dtype: object

In [32]:
SteamGames['genres'][0]

['Action', 'Casual', 'Indie', 'Simulation', 'Strategy']

In [33]:
SteamGames = SteamGames.explode("genres")
SteamGames = SteamGames.dropna(subset=["genres"])
SteamGames

publisher      genres             app_name                title   
0         Kotoshiro      Action  Lost Summoner Kitty  Lost Summoner Kitty  \
0         Kotoshiro      Casual  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro       Indie  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro  Simulation  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro    Strategy  Lost Summoner Kitty  Lost Summoner Kitty   
...             ...         ...                  ...                  ...   
32132  Laush Studio       Indie        Russian Roads        Russian Roads   
32132  Laush Studio      Racing        Russian Roads        Russian Roads   
32132  Laush Studio  Simulation        Russian Roads        Russian Roads   
32133      SIXNAILS      Casual  EXIT 2 - Directions  EXIT 2 - Directions   
32133      SIXNAILS       Indie  EXIT 2 - Directions  EXIT 2 - Directions   

                                                     url release_date   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
...                                                  ...          ...   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   

                                                    tags   
0          [Strategy, Action, Indie, Casual, Simulation]  \
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
...                                                  ...   
32132                        [Indie, Simulation, Racing]   
32132                        [Indie, Simulation, Racing]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
...                                                  ...             ...   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32133  http://steamcommunity.com/app/658870/reviews/?...             NaN   
32133  http://steamcommunity.com/app/658870/reviews/?...             NaN   

                                                   specs  price early_access   
0                                        [Single-player]   4.99        False  \
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
...                                                  ...    ...          ...   
32132  [Single

In [34]:
SteamGames['metascore'] = SteamGames['metascore'].replace('NA', np.nan)

**CARGA DEL CONJUNTO DE DATOS**

In [35]:
Save = 'data/user_games_clean.csv'
SteamGames.to_csv(Save, index=False, encoding='utf-8')

In [ ]:
#Indico donde quiero guardar el parquet y con que nombre
output_file= 'data/user_games_clean.parquet'

#Transformo a traves de una tabla el archivo csv en parquet
table = pa.Table.from_pandas(SteamGames)
pq.write_table(table,output_file)